### PR TITLE
feat: always make an HTTP call when calling self API methods

### PR DESCRIPTION
**Pull request by frostming**
Signed-off-by: Frost Ming <me@frostming.com>

## What does this PR address?

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/_bentoml_impl/server/app.py [hunk 1]
```diff
@@ -271,9 +271,20 @@ def server_request_hook(span: Span | None, _scope: dict[str, t.Any]) -> None:
...
+        from ..client import RemoteProxy
+
         self._service_instance = self.service()
         logger.info("Service %s initialized", self.service.name)
+        if deployment_url := os.getenv("BENTOCLOUD_DEPLOYMENT_URL"):
+            proxy = RemoteProxy(
+                deployment_url, service=self.service, media_type="application/json"
+            )
+        else:
+            proxy = RemoteProxy("http://localhost:3000", service=self.service, app=app)
```
Comment (aarnphm) [inline src/_bentoml_impl/server/app.py:284]: hmm this won't work if users specific custom ports right?

File: src/_bentoml_sdk/method.py [hunk 2]
```diff
@@ -107,12 +107,22 @@ def __get__(self: T, instance: None, owner: type) -> T: ...
...
+
+        if proxy := getattr(instance, "__self_proxy__", None):
+            func = getattr(proxy, self.name)
+        else:
+            func = local_caller
+        func.local = local_caller  # type: ignore[attr-defined]
+        return func
+
+    def _local_call(self, instance: t.Any) -> t.Callable[P, R]:
+        from pydantic.fields import FieldInfo
```
Comment (aarnphm) [inline src/_bentoml_sdk/method.py:123]: Why the import here? I assume this is a side effect in L113?

**Review by aarnphm — COMMENTED** [review]
what is the reason for setting the self API to be WSGI?

what happens if users want to use async call?

File: src/_bentoml_sdk/method.py (hunk 2 above)
Comment (frostming) [reply src/_bentoml_sdk/method.py:123]: it's used by L137

File: src/_bentoml_impl/client/http.py [hunk 3]
```diff
@@ -94,6 +96,13 @@ def _make_client(
             url = "http://127.0.0.1:3000"
         elif parsed.scheme == "tcp":
             url = f"http://{parsed.netloc}"
+        elif app is not None:
+            if client_cls is httpx.Client:
+                from a2wsgi import ASGIMiddleware
+
+                transport = httpx.WSGITransport(app=ASGIMiddleware(app))
```
Comment (frostming) [inline src/_bentoml_impl/client/http.py:103]: httpx.Client can only interact with WSGITransport while httpx.AsyncClient interacts with ASGITransport.
It depends on whether user uses sync or async client. @aarnphm

File: src/_bentoml_impl/server/app.py (hunk 1 above)
Comment (aarnphm) [reply src/_bentoml_impl/server/app.py:284]: ah gotcha

File: src/_bentoml_sdk/method.py (hunk 2 above)
Comment (aarnphm) [reply src/_bentoml_sdk/method.py:123]: oh i see.

**Review by aarnphm — APPROVED** [review]

**Pull request merged**